src/sqlfluff/rules/structure/ST06.py

```diff
@@ -86,35 +86,34 @@ def _eval(self, context: RuleContext) -> EvalResultType:
         ]  # type: ignore
 
         assert context.segment.is_type("select_clause")
-        # Ignore select clauses which belong to:
-        # - set expression, which is most commonly a union
-        # - insert_statement
-        # - create table statement
-        #
-        # In each of these contexts, the order of columns in a select should
-        # be preserved.
-        if len(context.parent_stack) >= 2 and context.parent_stack[-2].is_type(
-            "insert_statement", "set_expression"
-        ):
-            return None
-        if (
-            len(context.parent_stack) >= 3
-            and context.parent_stack[-3].is_type("insert_statement", "set_expression")
-            and context.parent_stack[-2].is_type("with_compound_statement")
-        ):
-            return None
-        if len(context.parent_stack) >= 3 and context.parent_stack[-3].is_type(
-            "create_table_statement", "merge_statement"
-        ):
-            return None
-        if (
-            len(context.parent_stack) >= 4
-            and context.parent_stack[-4].is_type(
-                "create_table_statement", "merge_statement"
-            )
-            and context.parent_stack[-2].is_type("with_compound_statement")
-        ):
-            return None
+
+        # insert, merge, create table, union are order-sensitive
+        for seg in reversed(context.parent_stack):
+            if seg.is_type(
+                "insert_statement",
+                "set_expression",
+                "create_table_statement",
+                "merge_statement",
+            ):
+                return None
+
+        # CTE is order-sensitive only if CTE is referenced in set expression
```

Comment:
Perhaps a small thing, which you can choose whether to address: The order is only sensitive _if a wildcard (*) is used_. In the case of a set expression where all the columns are re-specified, it would be ok to reorder them here - it's only when the set expression uses a `*` that it becomes a problem.

For bonus points, taking that into account would be awesome.